File: src/workflows/workflowGraphEngine.ts
/**
 * Converted from Python: core/workflows/workflow_graph_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function buildWorkflowGraph(objective: any, plan: any, state: any, execution: any, transitions: any): any {
  var nodes: any[] = [];
  var edges: any[] = [];
  var obj_name: any = py.toStr(py.get(objective, "objective", "objective"));
  py.listAppend(nodes, {"id": `objective:${py.toStr(obj_name)}`, "type": "objective"});
  var step: any;
  for (step of py.iter(py.slice(py.get(plan, "steps", []), null, 10000))) {
    var step_id: any = py.toStr(py.get(step, "id", ""));
    py.listAppend(nodes, {"id": step_id, "type": "step", "runtime": py.toStr(py.get(step, "runtime", ""))});
    var depends_on: any = py.toStr(py.get(step, "depends_on", ""));
    if (py.truthy(depends_on)) {
      py.listAppend(edges, {"from": depends_on, "to": step_id, "relation": "depends_on"});
    }
    py.listAppend(edges, {"from": `objective:${py.toStr(obj_name)}`, "to": step_id, "relation": "executes"});
  }
  var transition: any;
  for (transition of py.iter(py.slice(transitions, null, 10000))) {
    py.listAppend(edges, {"from": py.toStr(py.get(transition, "from", "")), "to": py.toStr(py.get(transition, "to", "")), "relation": "transitions"});
  }
  if ((py.get(state, "retries", 0) > 0)) {
    py.listAppend(nodes, {"id": "checkpoint:recovery", "type": "checkpoint"});
    py.listAppend(edges, {"from": "checkpoint:recovery", "to": `objective:${py.toStr(obj_name)}`, "relation": "recovers"});
  }
  py.listAppend(nodes, {"id": "semantic:state", "type": "semantic_state"});
  return {"nodes": py.sorted(nodes, {key: ((item: any) => py.at(item, "id")) as (item: any) => any}), "edges": py.sorted(edges, {key: ((item: any) => [py.get(item, "from", ""), py.get(item, "to", ""), py.get(item, "relation", "")]) as (item: any) => any}), "bounded": true};
}

File: lib/src/workflows/workflow_graph_engine.dart
// Port of core/workflows/workflow_graph_engine.py

Map<String, dynamic> buildWorkflowGraph(
  Map<String, dynamic> objective,
  Map<String, dynamic> plan,
  Map<String, dynamic> state,
  Map<String, dynamic> execution,
  List<Map<String, dynamic>> transitions,
) {
  final List<Map<String, dynamic>> nodes = <Map<String, dynamic>>[];
  final List<Map<String, dynamic>> edges = <Map<String, dynamic>>[];

  final String objName = '${objective['objective'] ?? 'objective'}';
  nodes.add(<String, dynamic>{'id': 'objective:$objName', 'type': 'objective'});

  final List<dynamic> steps = plan['steps'] is List
      ? List<dynamic>.from(plan['steps'] as List<dynamic>)
      : <dynamic>[];
  for (final dynamic stepRaw in steps.take(10000)) {
    final Map<String, dynamic> step = stepRaw is Map
        ? Map<String, dynamic>.from(stepRaw)
        : <String, dynamic>{};
    final String stepId = '${step['id'] ?? ''}';
    nodes.add(<String, dynamic>{
      'id': stepId,
      'type': 'step',
      'runtime': '${step['runtime'] ?? ''}',
    });
    final String dependsOn = '${step['depends_on'] ?? ''}';
    if (dependsOn.isNotEmpty) {
      edges.add(<String, dynamic>{
        'from': dependsOn,
        'to': stepId,
        'relation': 'depends_on',
      });
    }
    edges.add(<String, dynamic>{
      'from': 'objective:$objName',
      'to': stepId,
      'relation': 'executes',
    });
  }

  for (final Map<String, dynamic> transition in transitions.take(10000)) {
    edges.add(<String, dynamic>{
      'from': '${transition['from'] ?? ''}',
      'to': '${transition['to'] ?? ''}',
      'relation': 'transitions',
    });
  }

  final int retries = state['retries'] is int ? state['retries'] as int : 0;
  if (retries > 0) {
    nodes.add(
        <String, dynamic>{'id': 'checkpoint:recovery', 'type': 'checkpoint'});
    edges.add(<String, dynamic>{
      'from': 'checkpoint:recovery',
      'to': 'objective:$objName',
      'relation': 'recovers',
    });
  }

  nodes.add(<String, dynamic>{
    'id': 'semantic:state',
    'type': 'semantic_state',
  });

  nodes.sort((Map<String, dynamic> a, Map<String, dynamic> b) =>
      '${a['id']}'.compareTo('${b['id']}'));
  edges.sort((Map<String, dynamic> a, Map<String, dynamic> b) {
    final int c1 = '${a['from'] ?? ''}'.compareTo('${b['from'] ?? ''}');
    if (c1 != 0) {
      return c1;
    }
    final int c2 = '${a['to'] ?? ''}'.compareTo('${b['to'] ?? ''}');
    if (c2 != 0) {
      return c2;
    }
    return '${a['relation'] ?? ''}'.compareTo('${b['relation'] ?? ''}');
  });

  return <String, dynamic>{
    'nodes': nodes,
    'edges': edges,
    'bounded': true,
  };
}
